components/server/ee/src/workspace/gitpod-server-impl.ts

```diff
@@ -2149,48 +2150,60 @@ export class GitpodServerEEImpl extends GitpodServerImpl {
         return result;
     }
 
-    async listBilledUsage(ctx: TraceContext, req: BillableSessionRequest): Promise<ExtendedBillableSession[]> {
-        const { attributionId, startedTimeOrder, from, to } = req;
+    async listBilledUsage(ctx: TraceContext, req: ListBilledUsageRequest): Promise<ListBilledUsageResponse> {
+        const { attributionId, fromDate, toDate, perPage, page } = req;
         traceAPIParams(ctx, { attributionId });
         let timestampFrom;
         let timestampTo;
         const user = this.checkAndBlockUser("listBilledUsage");
 
         await this.guardCostCenterAccess(ctx, user.id, attributionId, "get");
 
-        if (from) {
-            timestampFrom = Timestamp.fromDate(new Date(from));
+        if (fromDate) {
+            timestampFrom = Timestamp.fromDate(new Date(fromDate));
         }
-        if (to) {
-            timestampTo = Timestamp.fromDate(new Date(to));
+        if (toDate) {
+            timestampTo = Timestamp.fromDate(new Date(toDate));
         }
         const usageClient = this.usageServiceClientProvider.getDefault();
         const response = await usageClient.listBilledUsage(
             ctx,
             attributionId,
-            startedTimeOrder as number,
+            ListBilledUsage.Ordering.ORDERING_DESCENDING,
+            perPage,
+            page,
             timestampFrom,
             timestampTo,
         );
-        const sessions = response.getSessionsList().map((s) => UsageService.mapBilledSession(s));
-        const extendedSessions = await Promise.all(
-            sessions.map(async (session) => {
-                const ws = await this.workspaceDb.trace(ctx).findWorkspaceAndInstance(session.workspaceId);
-                let profile: User.Profile | undefined = undefined;
-                if (session.workspaceType === "regular" && session.userId) {
-                    const user = await this.userDB.findUserById(session.userId);
-                    if (user) {
-                        profile = User.getProfile(user);
+        const sessions = await Promise.all(
+            response
+                .getSessionsList()
+                .map((s) => UsageService.mapBilledSession(s))
+                .map(async (session) => {
+                    const ws = await this.workspaceDb.trace(ctx).findWorkspaceAndInstance(session.workspaceId);
+                    let profile: User.Profile | undefined = undefined;
+                    if (session.workspaceType === "regular" && session.userId) {
+                        // TODO add caching to void repeated loading of same profile details here
```

Comment:
I guess we all know that this is bad in terms of complexity. But for later I think we should add this data redundantly (as well as contextURL)to the actual usage entries instead of optimizing augmentation here.

Reply:
True. That's just a low-hanging perf optimization.